### PR TITLE
fix bn fuse vardesc and add model saver

### PR DESCRIPTION
**Pull request by tensor-tang**
test=develop

./paddle/fluid/inference/tests/api/test_analyzer_save_model --infer_model=yourpath

### CODE REVIEW

File: paddle/fluid/inference/tests/api/CMakeLists.txt [hunk 1]
```diff
@@ -86,6 +86,9 @@ inference_analysis_test(test_analyzer_small_dam SRCS analyzer_dam_tester.cc
         EXTRA_DEPS ${INFERENCE_EXTRA_DEPS}
         ARGS --infer_model=${DAM_SMALL_INSTALL_DIR}/model --infer_data=${DAM_SMALL_INSTALL_DIR}/data.txt --max_turn_num=1 SERIAL)
 
+# save model
+inference_analysis_api_test(test_analyzer_save_model ${DAM_SMALL_INSTALL_DIR} analyzer_save_model_tester.cc SERIAL)
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/CMakeLists.txt:90]: 这个和`analyzer_dam_tester.cc`以及`analyzer_resnet50_tester.cc`中`save_optim_model`部分有些重复。

**Comment by tensor-tang** [comment]
补充下背景，这里添加可执行是为了方便mobile端调用fluid pass优化，同时修复bn fuse中的var desc，解决paddle mobile load时挂掉的问题。

File: paddle/fluid/inference/tests/api/CMakeLists.txt (hunk 1 above)
Comment (tensor-tang) [reply paddle/fluid/inference/tests/api/CMakeLists.txt:90]: 我可以统一到helper

File: paddle/fluid/inference/tests/api/analyzer_save_model_tester.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  SetConfig(&cfg);
+  cfg.SetModel(FLAGS_infer_model + "/__model__", FLAGS_infer_model + "/param");
+  std::string optimModelPath = FLAGS_infer_model + "/saved_optim_model";
+  SaveOptimModel(&cfg, optimModelPath);
+
+  cfg.pass_builder()->ClearPasses();
+  int origin_num_ops = GetNumOps(cfg);
+  cfg.SetModel(optimModelPath + "/model", optimModelPath + "/params");
+  int fused_num_ops = GetNumOps(cfg);
+  CHECK_LE(fused_num_ops, origin_num_ops);
```
Comment (luotao1) [inline paddle/fluid/inference/tests/api/analyzer_save_model_tester.cc:45]: 需要把`CHECK_LE(fused_num_ops, origin_num_ops);`的功能也挪到每个单测`TEST(Analyzer_xxx, save_optim_model)`中么？即不需要`analyzer_save_model_tester.cc`?

Comment (tensor-tang) [reply paddle/fluid/inference/tests/api/analyzer_save_model_tester.cc:45]: 这个没太必要吧，目的就是为了加一个save model的可执行。
并且每个独立单侧里面会检查结果对比，需要依赖数据的准备，所以这里才不比较。

**Comment by tensor-tang** [comment]
遇到一次
```
[08:08:38] :	 [Step 1/1] 67/73 Test #608: test_calibration_mobilenetv1 ............***Failed   43.00 sec
```
http://ci.paddlepaddle.org/viewLog.html?buildId=91050&tab=buildLog&buildTypeId=Paddle_PrCi&logTab=tree&filter=all

**Comment by tensor-tang** [comment]
又遇到一次
``` 
68/73 Test #608: test_calibration_resnet50 ...............***Failed   64.05 sec
```
http://ci.paddlepaddle.org/viewLog.html?buildId=91088&tab=buildLog&buildTypeId=Paddle_PrCi&logTab=tail

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 3]
```diff
@@ -136,18 +136,21 @@ void ConvBNFusePass::ApplyImpl(ir::Graph* graph) const {
...
+    // Get batch norm bias
+    auto* bn_bias_tensor =
+        scope->FindVar(bn_bias->Name())->GetMutable<LoDTensor>();
+
     // Create eltwise_y (conv bias) variable
     VarDesc eltwise_y_in_desc(
         patterns::PDNodeName(name_scope_, "eltwise_y_in"));
+    eltwise_y_in_desc.SetShape(framework::vectorize(bn_bias_tensor->dims()));
+    eltwise_y_in_desc.SetDataType(bn_bias_tensor->type());
+    eltwise_y_in_desc.SetLoDLevel(0);
```
Comment (luotao1) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: Why do you need to add line146-148?

**Comment by tensor-tang** [comment]
@hong19860320 Please help review this PR

**Comment by luotao1** [comment]
> 同时修复bn fuse中的var desc，解决paddle mobile load时挂掉的问题。

Could you paste the error?

**Comment by tensor-tang** [comment]
The error is just core dump，when load models because of missing some var informations.
This thing fixed here is 
before:
```
...
  vars {
    name: "conv_bn_fuse/eltwise_y_in/65"
    type {
      type: LOD_TENSOR
    }
    persistable: true
  }
...
```
After:
```
...
  vars {
    name: "conv_bn_fuse/eltwise_y_in/65"
    type {
      type: LOD_TENSOR
      lod_tensor {
        tensor {
          data_type: FP32
          dims: 116
        }
        lod_level: 0
      }
    }
    persistable: true
  }
...
```

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 3 above)
Comment (tensor-tang) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: commented here https://github.com/PaddlePaddle/Paddle/pull/17143#issuecomment-487828409

Comment (luotao1) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: `eltwise_y_in_desc.SetLoDLevel(0);`->`eltwise_y_in_desc.SetLoDLevel(bn_bias_tensor->lod_level())`;?

Comment (tensor-tang) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: `bn_bias_tensor ` do not have `lod_level()`

but I can use bn_bias_var

Comment (tensor-tang) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc]: done

**Review by hong19860320 — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM

**Comment by tensor-tang** [comment]
遇到一次MAC CI random fail

http://ci.paddlepaddle.org/viewLog.html?buildId=91465&buildTypeId=PaddleMac_MacPrCi&tab=buildLog&branch_PaddleMac=pull%2F16684
248 - test_dist_simnet_bow

**Comment by tensor-tang** [comment]
python35 random fail

 	  6 - retry_allocator_test (OTHER_FAULT)
http://ci.paddlepaddle.org/viewLog.html?buildId=91469&buildTypeId=Paddle_PrCiPython35&tab=buildLog&branch_Paddle=pull%2F16684

**Pull request merged**